### PR TITLE
Release v2.0.0

### PR DESCRIPTION
**Pull request by novln**
Fix race conditions, high concurrency situations and change code architecture using separated modules.

### CODE REVIEW

File: drivers/store/memory/store.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	key = fmt.Sprintf("%s:%s", store.Prefix, key)
+	now := time.Now()
+
+	count, expiration := store.cache.Get(key, rate.Period)
+
+	lctx := store.getContextFromState(now, rate, expiration, count)
+	return lctx, nil
+}
+
+func (store *Store) getContextFromState(now time.Time, rate limiter.Rate,
```
Comment (thoas) [inline drivers/store/memory/store.go]: you can remove this method from the struct, it does not use ``store``

File: drivers/store/redis/store.go [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+		count, ttl, err := store.peekValue(rtx, key)
+		if err == nil {
+			return count, ttl, nil
+		}
+	}
+	return 0, 0, errors.New("retry limit exceeded")
+}
+
+// peekValue will retrieve the counter and its expiration for given key.
+func (store *Store) peekValue(rtx *libredis.Tx, key string) (int64, time.Duration, error) {
```
Comment (thoas) [inline drivers/store/redis/store.go]: same, it's an utility method so no need to use a struct method

File: drivers/store/redis/store.go [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+		created, err := store.setValue(rtx, key, expiration)
+		if err == nil {
+			return created, nil
+		}
+	}
+	return false, errors.New("retry limit exceeded")
+}
+
+// setValue will try to initialize a new counter if given key doesn't exists.
+func (store *Store) setValue(rtx *libredis.Tx, key string, expiration time.Duration) (bool, error) {
```
Comment (thoas) [inline drivers/store/redis/store.go]: same

File: drivers/store/redis/store.go [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+		// If ttl is negative and there is an error, do not retry an update.
+		if ttl < 0 {
+			return 0, 0, err
+		}
+	}
+	return 0, 0, errors.New("retry limit exceeded")
+}
+
+// updateValue will try to increment the counter identified by given key.
+func (store *Store) updateValue(rtx *libredis.Tx, key string, expiration time.Duration) (int64, time.Duration, error) {
```
Comment (thoas) [inline drivers/store/redis/store.go]: same

File: drivers/store/redis/store.go [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+
+	pong, err := cmd.Result()
+	if err != nil {
+		return false, errors.Wrap(err, "limiter: cannot ping redis server")
+	}
+
+	return (pong == "PONG"), nil
+}
+
+func (store *Store) getContextFromState(now time.Time, rate limiter.Rate,
```
Comment (thoas) [inline drivers/store/redis/store.go]: same

Comment (thoas) [reply drivers/store/redis/store.go]: it seems this method is the same exact code as ``memory.store. getContextFromState``, let's re-use it as a static utility

**Comment by thoas** [comment]
great job 👍

File: drivers/store/redis/store.go [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+		if err == nil {
+			return created, nil
+		}
+	}
+	return false, errors.New("retry limit exceeded")
+}
+
+// setValue will try to initialize a new counter if given key doesn't exists.
+func (store *Store) setValue(rtx *libredis.Tx, key string, expiration time.Duration) (bool, error) {
+	pipe := rtx.Pipeline()
```
Comment (thoas) [inline drivers/store/redis/store.go]: no need to use a pipeline here since there is only one transaction?

**Comment by novln** [comment]
@thoas Thank you for your feedback.

**Comment by thoas** [comment]
@novln 👏

**Pull request merged**